module/zfs/zio_checksum.c

```diff
@@ -146,6 +223,30 @@ zio_checksum_label_verifier(zio_cksum_t *zcp, uint64_t offset)
 }
 
 /*
+ * Calls the template init function of a checksum which supports context
+ * templates and installs the template into the spa_t.
+ */
+static void
+zio_checksum_template_init(enum zio_checksum checksum, spa_t *spa)
+{
+	zio_checksum_info_t *ci = &zio_checksum_table[checksum];
+
+	if (ci->ci_tmpl_init == NULL)
+		return;
+	if (spa->spa_cksum_tmpls[checksum] != NULL)
```

Comment:
Good question.  I don't see why this couldn't/shouldn't be done at spa_init() time.  However, since this a port of an existing illumos patch I think we should bring over the code as is, unless it's clearly wrong, and then followup with optimizations.  That will make it easier to submit this stuff back upstream.